src/frontend/Ast_to_Mir.ml

```diff
@@ -582,35 +637,37 @@ let trans_prog filename
         (trans_stmt
            {dread= None; dconstrain= Some Check; dadlevel= AutoDiffable})
         transformedparametersblock
-  in
-  let modelb =
-    map
-      (trans_stmt {dread= None; dconstrain= None; dadlevel= AutoDiffable})
-      modelblock
-  in
-  let log_prob =
-    prepare_params
     @
     match modelb with
     | [] -> []
     | hd :: _ -> [{stmt= Block modelb; smeta= hd.smeta}]
   in
+  let gen_from_block declc block =
+    map (trans_stmt declc) (get_block block p) @ gen_writes block output_vars
+  in
   let generate_quantities =
-    prepare_params
-    @ map
-        (trans_stmt {dread= None; dconstrain= Some Check; dadlevel= DataOnly})
-        generatedquantitiesblock
+    let gen_decl_c = {dread= None; dconstrain= None; dadlevel= DataOnly} in
+    gen_from_block
+      {gen_decl_c with dread= Some ReadParam; dconstrain= Some Constrain}
+      Parameters
+    @ compiler_if "emit_transformed_parameters__"
+        (gen_from_block
+           {gen_decl_c with dconstrain= Some Check}
+           TransformedParameters)
+    @ compiler_if "emit_generated_quantities__"
+        (gen_from_block
+           {gen_decl_c with dconstrain= Some Check}
+           GeneratedQuantities)
   in
   let transform_inits =
     map
       (trans_stmt
-         { dread= Some ReadParam
+         { dread= Some ReadData
          ; dconstrain= Some Unconstrain
          ; dadlevel= DataOnly })
       parametersblock
   in
   { functions_block=
-      (* Should this be AutoDiffable for functions here?*)
       Option.value_map functionblock ~default:[] ~f:(fun fundefs ->
           List.map fundefs ~f:(fun fundef ->
               trans_fun_def
```

Comment:
Is it important that FunDef's get translated as AutoDiffable below here?

Reply:
Functions should be generally AutoDiffable, though one day we can make use of the autodiff level on expressions in the optimization pass to do that more optimally ;)